Develop --> master | backend

Yaroslav7125 opened the pull request.

Yaroslav7125 committed:
init proj

Yaroslav7125 committed:
add .name

Yaroslav7125 committed:
Add sqlize

Yaroslav7125 committed:
Add requests

Yaroslav7125 committed:
Add refs for requests

Yaroslav7125 committed:
Add nuxt front

Yaroslav7125 committed:
Rewrote rename title todo

Yaroslav7125 committed:
Add working front and working back

Yaroslav7125 committed:
Rewrote changing completed.

Yaroslav7125 committed:
Add migrations.

Yaroslav7125 committed:
Add models.

Yaroslav7125 committed:
Fix bug with first todo in todoes.

Yaroslav7125 committed:
Refactor comments.

Yaroslav7125 committed:
Add seeds.

crutch12 commented:
Блин.. ну зачем ты сюда копипастнул nuxt проект
Переделал бы старый, в нём бы создал папку front
Мне заново весь front смотреть?)

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+let dbTodos = todos(sequelize, DataTypes);
+
+(async ()=>{
+    await sequelize.sync();//{ force: true }
```
Нафига sync, если у тебя теперь есть мигарции?
Создай отдельный скрипт в package.json, который проливает миграции
И эту команду нужно будет вызывать перед стартом приложения через docker-compose (например)

crutch12 replied:
А пока нету докера: ну просто руками вызывать migrate перед стартом

Yaroslav7125 replied:
Исправил в коммите a676435.

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+koa.use(cors());
+
+/// db connecting
+const sequelize = new Sequelize('TodoDB', 'postgres', '123', {
```
Логин/пароль хранить в коде - дикость конечно
Используй переменные среды
`process.env.DB_LOGIN || 'postgres'`
`process.env.DB_PASSWORD || '123'`

А сами переменные можешь задавать через файлик .env, или модуль cross-env, или через docker-compose в будущем

Просто сейчас девопсер не сможет указать свой логин пароль, т.к. ты не дал такую возможность

crutch12 replied:
Вот пример
https://github.com/santoslos/todoListWithAPI/blob/eb6b0fab35db25c12f5b244b982a88748d5b8567/ormconfig.js#L3

Yaroslav7125 replied:
Исправил в коммите a676435.

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+    let theData = await dbTodos.findAll({
+        attributes: ['id', 'title', 'completed'],
+    });
+    let todos = [];
```
А нафига тебе промежуточный массив?

Yaroslav7125 replied:
Исправил в коммите 380a258.

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+        attributes: ['id', 'title', 'completed'],
+    });
+    let todos = [];
+    theData.forEach((elm)=>{
```
Ну и такие вещи делаются через `.map`

Yaroslav7125 replied:
Исправил в коммите 380a258.

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+
+async function addData(todo){ // принимает обьект task с полями title, completed
+    if(todo.title){
+        let addedTodo = await dbTodos.create({title:`${todo.title}`, completed:`${todo.completed}`});
```
Почему у тебя completed отправляется строкой? это же boolean

Yaroslav7125 replied:
Исправил в коммите 69e8b8d.

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+async function addData(todo){ // принимает обьект task с полями title, completed
+    if(todo.title){
+        let addedTodo = await dbTodos.create({title:`${todo.title}`, completed:`${todo.completed}`});
+        return addedTodo.id;
```
А че весь todo не вернуть?

Yaroslav7125 replied:
Исправил в коммите bd3812c.

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+
+async function deleteData(delId){ // принимает id элемента который следует удалить
+    await dbTodos.destroy({
+        attributes: ['id', 'title', 'completed'],
```
А зачем attributes здесь?

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+};
+
+async function getDataById(dataId){ // принимает id элемента - который вернёт
+    let theData = await dbTodos.findAll({
```
`dbTodos.findById` вроде есть
Ты сейчас массив возвращаешь, хотя правильнее 1 элемент / либо null

Yaroslav7125 replied:
Удалил getDataById.

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+async function changeCompleted(dataId, complFlag){//принимает id элемента который нужно изменить
+    getDataById(dataId).then((data)=>{
+        const todo = data[0].dataValues;
+        (async ()=>{
```
Чего?)
Я вообще не понял зачем тебе `getDataById` здесь, но допустим. Тогда это должно выглядеть так:

```
const found = await getDataById(dataId)
const todo = found[0].dataValues;
return dbTodos.update(...)
```

crutch12 replied:
Либо так
```
return getDataById(dataId).then((data)=>{
  ...
  return dbTodos.update(...)
})
```

crutch12 replied:
Но я бы так написал:
```
return await dbTodos.update(...)
```
И не ебал себе мозг, т.к. sequelize сам проверит на существование

Yaroslav7125 replied:
Исправил в коммите 3057d19

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+            ctx.body = (todos);
+        } else {
+            ctx.status = 204;
+            ctx.body = 'No content';
```
Ну... что мешает вернуть `[]` вместо это?)
Наверное очень удобно на клиенте по одной и той же апишке получать то json, то строку?)))

Yaroslav7125 replied:
Поправил

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+        }
+})
+    .get('/tasks/:id', async (ctx)=>{  // принимает id таски которую возвращает        READ
+        let task = await getDataById(ctx.params.id);
```
Реализация getDataById не очень, писал сверху

Yaroslav7125 replied:
getDataByid больше не использую.

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+router.post('/tasks', async (ctx)=>{  // принимает обьект таски которую сетит в бд      CREATE
+    let newTodo = (ctx.request.body);
+    newTodo.id = await addData(newTodo);
+    ctx.response.body = newTodo;
```
Переделай addData и верни то, что вернёт он :)

Yaroslav7125 replied:
Переделал

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+    await sequelize.sync();//{ force: true }
+})();
+
+async function getData(){
```
Переименуй все `Data` на `Todo` или `Task`, нафига абстрактные назщвания, если ты работаешь с таблицей todos?

Yaroslav7125 replied:
Переименовал в коммите 380a258

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+    ctx.response.body = newTodo;
+});
+
+router.put('/tasks/change-title/:id/:newTitle', async (ctx)=>{ // принимает id и новый title     UPDATE
```
REST через :params указывает доступ к ресурсу, но никак не новое значение!
`newTitle` надо передавать через body запроса, у тебя же put

Yaroslav7125 replied:
Исправил, получаю переданное значение в теле запроса.

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+   ctx.body = await changeTitleTodo(ctx.params.id, ctx.params.newTitle);
+});
+
+router.put('/tasks/change-completed/:id/:completFlag', async (ctx)=>{ // принимает id и меняет completed у соответвующей таски    UPDATE
```
То же самое, что сверху
В целом там можно делать, но это не очень классно

Yaroslav7125 replied:
Исправил, получаю переданное значение в теле запроса.

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,125 @@
+    ctx.status = 202;
+});
+
+router.delete('/tasks/:id', async (ctx)=>{ // принимает id на удаление       DELETE
```
Ну хоть тут всё ок)

crutch12 commented on backend/models/todoes.js:
```diff
@@ -0,0 +1,24 @@
+  'use strict';
```
todoes.js? серьёзно?

Yaroslav7125 replied:
Добавил миграцию по созданию todosTable и удалению Todoes

Yaroslav7125 commented:
Посмотри только бэк, я потом в старый фронт его залью)

crutch12 reviewed (commented):
Написал комменты

Yaroslav7125 committed:
Убрал синхронизацию с базой данных, так-как есть миграции
Добавил использование переменных среды
Добавил скрипт по запуску миграций и сидов

Yaroslav7125 committed:
Переименовал все data на todo.
getData/getTodos - функция получения всех todo из базы данных:
 променял обрезку данных на .map
 удалил использование промежуточного массива.

Yaroslav7125 committed:
Исправил отправку completed со строки на boolean

Yaroslav7125 committed:
Подправил добавление todo в базу.
Теперь с бэка возвращается вся todo.

Yaroslav7125 committed:
Переписал функцию changeCompleted. Теперь функция возвращает todo, completed которого был ищменён. Также удалил использование функции getTodosById.

Yaroslav7125 committed:
Удалил getTodosById.

Yaroslav7125 committed:
Изменил возврат тудушек - когда их нет вернётся пустой массив.

Yaroslav7125 committed:
Изменил смену title - title приходит не параметром в строке а в теле запроса

crutch12 commented on backend/app.js:
```diff
@@ -0,0 +1,99 @@
+})
+    .put('/tasks/change-completed/:id/:completFlag', async (ctx)=>{ // принимает id и меняет completed у соответвующей таски    UPDATE
+    await changeCompleted(ctx.params.id, ctx.params.completFlag);
+    ctx.body = 'Accepted';
```
Очень удобно, наверное, делать запрос на обновление сущности и получать в ответ какую-то строку
Вместо того, чтобы получить изменённую сущность и с ней дальше работать?

Yaroslav7125 replied:
Исправил в коммите fa51b67.

Yaroslav7125 committed:
Переписал смену completed - параметром передаеётся только id - само значение в теле запроса

Yaroslav7125 committed:
Переписал изменение title - запрос вернёт изменённую todo

Yaroslav7125 committed:
Переправил - changeTitleTodo - тпереь при успешном изменении вернёться имененная todo

Yaroslav7125 committed:
Подправил getTodo - убрал оатрибуты,
changeCompleted - убрал использование .findByPk и добавил использвание returning.
changeTitleTodo - сделал тоже самое то и в changeCompleted.

Yaroslav7125 committed:
Подправил комменты к коду и добавил host и dialect в переменные среды

Yaroslav7125 committed:
Добавил миграцию
Переименовал сиды
Подправил имня модели

Yaroslav7125 committed:
Добавил миграцию которая удаляет Todoes.

Yaroslav7125 committed:
Удалил front

Yaroslav7125 committed:
Добавил ts файл

Yaroslav7125 committed:
Добавил директории dist и src подправил запуск.

Yaroslav7125 committed:
Типизировал контекст ответа
изменил deleteTodo - теперь возвращает id удалённого todo

Yaroslav7125 committed:
Добавил типизацию для тела запросов
